### ISSUE TITLE
Feature idea - Fallback icon packs

### ISSUE THREAD

**Issue by comradekingu** [falzonv/discreet-launcher#130]
Some icon packs are nice, but have nowhere near good coverage of icons.
It would be nice to be able to select more than one preference.

If the default is

```
+
1.Default

```
and clicking it two times and selecting first "Articons Dark" and then "Frost" would produce:

```
+
1.  Frost [X]
2. Articons Dark [X)
3. Default
```

would work?

Or that selection thing from Favorites. Might be better, because the new page could have some sort of preview(?)

**Comment by falzonv** [comment]
Hello,

Thank you for the suggestion!

Unless adding a lot of code complexity (which I do not want), none of the two proposed options are compatible with the Android default Preference library.

If this feature is implemented, it will probably be as two settings:
- The existing icon pack setting (no change), maybe with a new name to suggest it is the main icon pack
- A new setting just below, exactly like the above, but with a name suggesting it is the fallback icon pack

Both settings would show the same options (that is, all installed icon packs + "none") and each app icon would be searched first in the main icon pack, then (if not found) in the fallback icon pack, then (if not found) it would use the default icon.

Although it would be good to make the choices of the fallback icon pack dynamic (to filter out the main icon pack as it changes), it doesn't seem to be possible without closing and re-opening the settings so it will not be done.

Therefore, if the user selects two times the same icon pack, the icon should be searched only once (if it was not found the first time, no need to search again in the same pack).
And if (s)he selects "none" as main icon pack, the fallback icon pack should be ignored.

Would it work for you?

Best regards.

**Comment by comradekingu** [comment]
That is what I had in mind. Small inelegant quirks are largely irrelevant.
The main idea is to give new and upcoming icon packs a chance, and I imagine Discreet launcher could be promoted by those with this setting :)

**Comment by falzonv** [comment]
Hello,

Since some time I work in parallel on both this ticket and #149 (as both are related to icons).
Good news, both are now almost finished!

*By the way, with this ticket in mind, I implemented the color filter of #149 to apply only on default icons.
That way it can be used to fill the remaining icon packs "holes" with a uniform appearance.*

---

Now I struggle a bit with some strings for this ticket and would like your advice:

As previously announced, there are now two settings.
They look exactly like the current icon pack setting and allow to select two icon packs.
Each app icon is searched in the icon pack from the first setting, then (if it was not found) in the icon pack from the second setting and then (if still not found) the default icon is used.

Is it clear enough if these settings are named as follow?
```
- Main icon pack
- Secondary icon pack
```

Also, the current icon pack setting is named `Use an installed icon pack` and I introduced "installed" after #15 to make it more clear for users not familiar with icon packs.
Since I couldn't find a proper way to insert this "installed" notion in the new settings names, I found a way to display a small popup message if the user opens the new icon pack settings with no icon packs currently installed.

For this message I have some ideas but am not sure of which (if any) is best:
- "You can install icon packs from your app store"
- "No icon packs found, did you installed some?"
- "Couldn't find any installed icon pack"

Any suggestions?

Of course, if you have an idea of settings names embedding the "installed" notion, I will be glad to hear it.
It would allow to keep the code simpler while preventing an unnecessary click for the user ;-)
*(We cannot use the setting subtitle line as it is already used to display the name of the selected icon pack.)*

Best regards.

**Comment by comradekingu** [comment]
I am thinking "Primary" and "Secondary" under an "Icon pack" header.

"Could not find any installed icon packs"
but better
"Install an icon pack to apply first"

"Preferred icon packs"
"Icon packs to use"

With "1. Arcticons Dark 2. Arcticons light" underneath?
and
"Install an icon pack to apply first"
in the subtext only if none are installed?

**Comment by falzonv** [comment]
Hello,

Thank you for your feedback!

> I am thinking "Primary" and "Secondary" under an "Icon pack" header.

=> I will go with this suggestion because it also gave me a better idea for the "no icon packs" message: instead of a small popup on opening the setting, it is now a text which is added in the settings when the user has no icon packs.

The screenshots below show how it looks like (I also disabled the settings when there is no icon packs).

For the text I also made a mix to say first that there is no icon packs, and then how the user can install some.
=> Does it looks good?

| With 2 icon packs selected | With no icon packs installed |
| --- | --- | 
| <img src="https://user-images.githubusercontent.com/4698531/145972532-3b59d963-d279-4ae2-b37d-28dab2041407.jpg"/> | <img src="https://user-images.githubusercontent.com/4698531/145977991-954103c9-0744-40e9-ad15-777e264777e6.jpg"/> |

---

> "Preferred icon packs"
"Icon packs to use"
With "1. Arcticons Dark 2. Arcticons light" underneath?
and
"Install an icon pack to apply first"
in the subtext only if none are installed?

=> As explained in the first comment, I prefer to stay with the Android default Preference library rather than create a customized preference type (this make things much easier to maintain).

Best regards.

**Comment by comradekingu** [comment]
Looks great. :) I think a \n for
"Could not find any icon packs\n(you can install them from your app store)."
makes sense.

**Comment by falzonv** [comment]
Hello,

As announced in the translations ticket, this has been published in new release v5.4.0!
I will leave this ticket open until it is available on F-Droid (usually takes 4 to 8 days).

Best regards.

**Comment by falzonv** [comment]
Since yesterday v5.4.0 is now available on F-Droid so I will close this ticket.
Feel free to add comments if needed or wished.
